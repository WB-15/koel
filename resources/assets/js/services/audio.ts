export const audio = {
  context: null as AudioContext | null,
  source: null as MediaElementAudioSourceNode | null,
  element: null as HTMLMediaElement | null,

  init (element: HTMLMediaElement): void {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext (): AudioContext {
    return this.context!
  },

  getSource (): MediaElementAudioSourceNode {
    return this.source!
  },

  getElement (): HTMLMediaElement {
    return this.element!
  }
}
