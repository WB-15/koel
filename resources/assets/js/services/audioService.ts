export const audioService = {
  context: null as unknown as AudioContext,
  source: null as unknown as MediaElementAudioSourceNode,
  element: null as unknown as HTMLMediaElement,

  init (element: HTMLMediaElement) {
    const AudioContext = window.AudioContext ||
      window.webkitAudioContext ||
      window.mozAudioContext ||
      window.oAudioContext ||
      window.msAudioContext

    this.context = new AudioContext()
    this.source = this.context.createMediaElementSource(element)
    this.element = element
  },

  getContext () {
    return this.context
  },

  getSource () {
    return this.source
  },

  getElement () {
    return this.element
  }
}
