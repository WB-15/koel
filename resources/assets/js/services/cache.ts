const CACHE_EXPIRATION_TIME = 1000 * 60 * 60 * 24 // 1 day

export const Cache = {
  storage: new Map<any, {
    time: number,
    value: any
  }>(),

  normalizeKey: (key: any) => typeof key === 'object' ? JSON.stringify(key) : key,

  has (key: any) {
    return this.hit(this.normalizeKey(key))
  },

  set (key: any, value: any) {
    this.storage.set(this.normalizeKey(key), {
      value,
      time: Date.now()
    })
  },

  hit (key: any) {
    return !this.miss(this.normalizeKey(key))
  },

  miss (key: any) {
    key = this.normalizeKey(key)

    if (!this.storage.has(key)) return true
    const { time } = this.storage.get(key)!

    if (time < Date.now() - CACHE_EXPIRATION_TIME) {
      this.storage.delete(key)
      return true
    }

    return false
  },

  invalidate (key: any) {
    this.storage.delete(this.normalizeKey(key))
  },

  resolve<T> (key: any, resolver: Closure) {
    key = this.normalizeKey(key)

    this.hit(key) || this.set(key, resolver())
    return this.storage.get(key)!.value
  }
}
