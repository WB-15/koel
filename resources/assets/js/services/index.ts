export * from './httpService'
export * from './downloadService'
export * from './localStorageService'
export * from './playbackService'
export * from './youTubeService'
export * from './socketService'
export * from './audioService'
export * from './uploadService'
export * from './authService'
