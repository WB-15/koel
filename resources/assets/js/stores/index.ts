export * from './albumStore'
export * from './artistStore'
export * from './equalizerStore'
export * from './favoriteStore'
export * from './playlistStore'
export * from './preferenceStore'
export * from './queueStore'
export * from './recentlyPlayedStore'
export * from './settingStore'
export * from './commonStore'
export * from './songStore'
export * from './userStore'
export * from './searchStore'
export * from './themeStore'
