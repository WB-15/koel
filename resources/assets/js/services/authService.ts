import { localStorageService } from '@/services'

const STORAGE_KEY = 'api-token'

export const authService = {
  getToken: () => localStorageService.get<string | null>(STORAGE_KEY),

  hasToken () {
    return Boolean(this.getToken())
  },

  setToken: (token: string) => localStorageService.set(STORAGE_KEY, token),
  destroy: () => localStorageService.remove(STORAGE_KEY)
}
