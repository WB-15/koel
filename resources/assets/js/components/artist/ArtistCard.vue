<template>
  <article
    v-if="showing"
    :class="layout"
    :title="artist.name"
    class="item"
    data-testid="artist-card"
    draggable="true"
    tabindex="0"
    @dblclick="shuffle"
    @dragstart="dragStart"
    @contextmenu.prevent="requestContextMenu"
  >
    <span class="thumbnail-wrapper">
      <ArtistThumbnail :entity="artist"/>
    </span>

    <footer>
      <div class="info">
        <a class="name" :href="`#!/artist/${artist.id}`" data-testid="name">{{ artist.name }}</a>
      </div>
      <p class="meta">
        <span class="left">
          {{ pluralize(artist.album_count, 'album') }}
          •
          {{ pluralize(artist.song_count, 'song') }}
          •
          {{ pluralize(artist.play_count, 'play') }}
        </span>
        <span class="right">
          <a
            :title="`Shuffle all songs by ${artist.name}`"
            class="shuffle-artist"
            href
            role="button"
            data-testid="shuffle-artist"
            @click.prevent="shuffle"
          >
            <i class="fa fa-random"/>
          </a>
          <a
            v-if="allowDownload"
            :title="`Download all songs by ${artist.name}`"
            class="download-artist"
            href
            role="button"
            data-testid="download-artist"
            @click.prevent="download"
          >
            <i class="fa fa-download"/>
          </a>
        </span>
      </p>
    </footer>
  </article>
</template>

<script lang="ts" setup>
import { computed, defineAsyncComponent, toRef, toRefs } from 'vue'
import { eventBus, pluralize, startDragging } from '@/utils'
import { artistStore, commonStore, songStore } from '@/stores'
import { downloadService, playbackService } from '@/services'

const ArtistThumbnail = defineAsyncComponent(() => import('@/components/ui/AlbumArtistThumbnail.vue'))

const props = withDefaults(defineProps<{ artist: Artist, layout?: ArtistAlbumCardLayout }>(), { layout: 'full' })
const { artist, layout } = toRefs(props)

const allowDownload = toRef(commonStore.state, 'allow_download')

const showing = computed(() => artist.value.song_count && !artistStore.isVarious(artist.value))

const shuffle = async () => {
  await playbackService.queueAndPlay(await songStore.fetchForArtist(artist.value), true /* shuffled */)
}

const download = () => downloadService.fromArtist(artist.value)
const dragStart = (event: DragEvent) => startDragging(event, artist.value, 'Artist')
const requestContextMenu = (event: MouseEvent) => eventBus.emit('ARTIST_CONTEXT_MENU_REQUESTED', event, artist.value)
</script>

<style lang="scss">
@include artist-album-card();
</style>
