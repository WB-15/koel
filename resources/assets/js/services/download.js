import $ from 'jquery';
import { map } from 'lodash';

import { playlistStore, favoriteStore } from '../stores';
import { ls } from '.';

export const download = {
    /**
     * Download individual song(s).
     *
     * @param {Array.<Object>|Object} songs
     */
    fromSongs(songs) {
        songs = [].concat(songs);
        const ids = map(songs, 'id');
        const params = $.param({ songs: ids });

        return this.trigger(`songs?${params}`);
    },

    /**
     * Download all songs in an album.
     *
     * @param {Object} album
     */
    fromAlbum(album) {
        return this.trigger(`album/${album.id}`);
    },

    /**
     * Download all songs performed by an artist.
     *
     * @param {Object} artist
     */
    fromArtist(artist) {
        // It's safe to assume an artist always has songs.
        // After all, what's an artist without her songs?
        // (See what I did there? Yes, I'm advocating for women's rights).
        return this.trigger(`artist/${artist.id}`);
    },

    /**
     * Download all songs in a playlist.
     *
     * @param {Object} playlist
     */
    fromPlaylist(playlist) {
        if (!playlistStore.getSongs(playlist).length) {
            console.warn('Empty playlist.');
            return;
        }

        return this.trigger(`playlist/${playlist.id}`);
    },

    /**
     * Download all favorite songs.
     */
    fromFavorites() {
        if (!favoriteStore.all.length) {
            console.warn("You don't like any song? Come on, don't be that grumpy.");
            return;
        }

        return this.trigger('favorites');
    },

    /**
     * Build a download link using a segment and trigger it.
     *
     * @param  {string} uri The uri segment, corresponding to the song(s),
     *                      artist, playlist, or album.
     */
    trigger(uri) {
        const sep = uri.indexOf('?') === -1 ? '?' : '&';
        const frameId = `downloader${Date.now()}`;
        $(`<iframe id="${frameId}" style="display:none"></iframe`).appendTo('body');
        document.getElementById(frameId).src = `/api/download/${uri}${sep}jwt-token=${ls.get('jwt-token')}`;
    },
}
