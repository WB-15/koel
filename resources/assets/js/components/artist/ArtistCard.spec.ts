import { mockHelper, render } from '@/__tests__/__helpers__'
import { cleanup, fireEvent } from '@testing-library/vue'
import { beforeEach, expect, it, vi } from 'vitest'
import factory from '@/__tests__/factory'
import { commonStore } from '@/stores'
import ArtistCard from './ArtistCard.vue'
import { downloadService, playbackService } from '@/services'

let artist: Artist

beforeEach(() => {
  vi.restoreAllMocks()
  mockHelper.restoreAllMocks()
  cleanup()

  artist = factory<Artist>('artist', {
    id: 3, // make sure it's not "Various Artists"
    name: 'Led Zeppelin',
    albums: factory<Album>('album', 4),
    songs: factory<Song>('song', 16)
  })

  commonStore.state.allowDownload = true
})

it('renders', () => {
  const { getByText, getByTestId } = render(ArtistCard, {
    props: {
      artist
    }
  })

  expect(getByTestId('name').innerText).equal('Led Zeppelin')
  getByText(/^4 albums\s+â€¢\s+16 songs.+0 plays$/)
  getByTestId('shuffle-artist')
  getByTestId('download-artist')
})

it('downloads', async () => {
  const mock = mockHelper.mock(downloadService, 'fromArtist')

  const { getByTestId } = render(ArtistCard, {
    props: {
      artist
    }
  })

  await fireEvent.click(getByTestId('download-artist'))
  expect(mock).toHaveBeenCalledTimes(1)
})

it('shuffles', async () => {
  const mock = mockHelper.mock(playbackService, 'playAllByArtist')

  const { getByTestId } = render(ArtistCard, {
    props: {
      artist
    }
  })

  await fireEvent.click(getByTestId('shuffle-artist'))
  expect(mock).toHaveBeenCalled()
})
