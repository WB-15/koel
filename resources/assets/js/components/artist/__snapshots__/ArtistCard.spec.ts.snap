// Vitest Snapshot v1

exports[`renders 1`] = `
<article data-v-f01bdc56="" class="item full" draggable="true" tabindex="0" title="Led Zeppelin"><br data-v-f01bdc56="" data-testid="thumbnail" entity="[object Object]">
  <footer data-v-f01bdc56="">
    <div data-v-f01bdc56="" class="name"><a href="#/artist/42" class="text-normal" data-testid="name">Led Zeppelin</a></div>
    <p data-v-f01bdc56="" class="meta"><a title="Shuffle all songs by Led Zeppelin" class="shuffle-artist" role="button"> Shuffle </a><a title="Download all songs by Led Zeppelin" class="download-artist" role="button"> Download </a></p>
  </footer>
</article>
`;
